[Patch 2/2] Feature/client capability response [Test 4.6.2]

CoralMalachi opened the pull request:
### Description
According to Multi-AP Specification, If a Multi-AP Agent receives a
Client Capability Query message, then within one second it shall respond
with a Client Capability Report message.

If the STA specified in the Client Capability Query message is not
associated with any of the BSS operated by the Multi-AP Agent (an error
scenario), the Multi-AP Agent shall set the result code in the Client
Capability Report TLV to 0x01 and include an Error Code TLV.
If it is not an error scenario, the Client Capability Report TLV should
include the frame body of the most recently received (Re)Association
Request frame from this client.

### Changes

- Parse the association request frame from the AP-STA-CONNECTED event.

- Save in the associated_clients_t map the frame body of the most recently 
received (Re)Association Request frame from each client.

- Add  the frame body of the most recently received (Re)Association
Request frame from the client to the Client Capability Report TLV

### Testbed status
Passed, you can check out the logs:https://gitlab.com/prpl-foundation/prplMesh/pipelines/133224675

there is a bug in the master version #1097, the logs above are before rebasing with the master which causes the bug---solved !

latest logs : https://gitlab.com/prpl-foundation/prplMesh/pipelines/135801475

### Next steps

- [x] Add an association frame in dummy mode.
- [x] Add a test to check it [dummy mode].
- [x] Currently, I'm facing a strange issue that the association frame is not delivered from son slave to backhaul manager, need to debug it. -----> Solved! an issue with the TLVF, thanks to @LiorAmram 
- [x] Solve the issue on the testbed

<img src="https://media.giphy.com/media/AUGqun7QALo0o/giphy.gif" width=144>
Signed-off-by: Coral Malachi coral.malachi@intel.com

Commit:
tlvf: client caps report: update association frame to unkown length list

According to Multi-AP Specification, the Client Capability Report TLV does
not contain a length value for the frame body of the most recently
received (Re)Association Request frame from the same client.

Therefore, remove the length attribute and make the `association_frame`
attribute to an unknown length list.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: update client caps report TLV auto-generated files

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
agent: ap_manager: send the association frame to son slave

Preparative commit.

Send the association frame to the son slave as part of the
cACTION_APMANAGER_CLIENT_ASSOCIATED_NOTIFICATION vendor-specific message.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: beerocks message backhaul: add association frame

Preparative commit.

When the AP manager received a vendor-specific message of type
ACTION_APMANAGER_CLIENT_ASSOCIATED_NOTIFICATION, a vendor-specific message
of type `cACTION_BACKHAUL_CLIENT_ASSOCIATED_NOTIFICATION` is created and
being sent to the backhaul manager.
In order to add the association frame to this message, we need to add a
new field to the cACTION_BACKHAUL_CLIENT_ASSOCIATED_NOTIFICATION of type
string.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: beerocks message backhaul: update TLVs auto-generated files

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
agent: son slave: send the association frame to backhaul manager

Send the association frame to the backhaul manager as part of the
cACTION_BACKHAUL_CLIENT_ASSOCIATED_NOTIFICATION vendor-specific message.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
bwl: dummy: add dummy association frame

Preparative commit.
preparation for a test - send a dummy association frame from
`ap_wlan_hal_dummy` to `ap_manager` in order to test it in a future
commit.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

tomereli commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -1812,15 +1812,16 @@ bool ap_wlan_hal_dwpal::process_dwpal_event(char *buffer, int bufLen, const std:
         memset(msg_buff.get(), 0, sizeof(sACTION_APMANAGER_CLIENT_ASSOCIATED_NOTIFICATION));
         memset((char *)&msg->params.capabilities, 0, sizeof(msg->params.capabilities));
 
-        char VAP[SSID_MAX_SIZE]        = {0};
```
Commit message is wrong - we need the (re)association frame, not the (re)authentication frame.

tomereli commented on common/beerocks/bwl/include/bwl/base_wlan_hal_types.h:
```diff
@@ -253,6 +253,7 @@ typedef struct {
 } sHOSTAP_ENABLED_NOTIFICATION;
 
 #define SSID_MAX_SIZE beerocks::message::WIFI_SSID_MAX_LENGTH
+#define ASSOCIATION_FRAME_SIZE 1024
```
This *has* to be verified in hostapd source code that this is indeed the maximum size of the association frame, otherwise, we are risking truncating the data here.

arnout replied:
I believe the absolute maximum (according to 802.11 specification, not hostapd) would be the MMPDU size of 2304 bytes. I think somebody with actual WiFi knowledge should be able to answer that easily.

tomereli replied:
According to a WiFi expert (WAVE600 FW team leader) the size of the assoc_req is a couple of hundreds of bytes at the most. It is true that the MMPDU is the absolute maximum, so we can set it to a nice round number like 2048 bytes and add a comment saying why we chose this number.

tomereli commented on common/beerocks/bwl/include/bwl/base_wlan_hal_types.h:
```diff
@@ -190,6 +190,7 @@ typedef struct {
     uint8_t reserved1;
     uint8_t reserved2;
     uint8_t reserved3;
+    char *association_frame;
```
This is not OK, why not use `std::string`? and if it is a `char *`, then also add `size_t association_frame_len`.
Also please change to `uint8_t` so it is aligned to the rest of the file.

arnout replied:
We shouldn't use std::string for a byte array. I would like us to start using `std::basic_string<uint8_t>`, but that would be the first time we use it in prplMesh so let's first get convergence on it.

tomereli commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -1902,6 +1902,8 @@ bool ap_wlan_hal_dwpal::process_dwpal_event(char *buffer, int bufLen, const std:
         msg->params.vap_id       = beerocks::utils::get_ids_from_iface_string(VAP).vap_id;
         msg->params.mac          = beerocks::net::network_utils::mac_from_string(MACAddress);
         msg->params.capabilities = {};
+        std::copy_n(assoc_req, strnlen(assoc_req, ASSOCIATION_FRAME_SIZE) + 1,
```
What's the `+ 1` for? is it for the null terminator? We need to verify in the MultiAP specification if the request should be null terminated or not. As a general rule, all strings are not null terminated in TLVs, take the manufacturer string in the M2 TLV as an example.

morantr replied:
I think `strnlen` is one of the forbidden SDL functions, no?
@tomereli

tomereli replied:
Actually it is not, `strlen` is forbidden - see https://github.com/intel/safestringlib/wiki/SDL-List-of-Banned-Functions

CoralMalachi replied:
I have added the +1 for the null terminator to get a result from the `strnlen` method, but when I'm filling the tlvClientCapabilityReport I copy it without the null terminator. @tomereli

tomereli commented on framework/tlvf/yaml/tlvf/wfa_map/tlvClientCapabilityReport.yaml:
```diff
@@ -10,12 +10,9 @@ tlvClientCapabilityReport:
   association_frame:
-    _type: uint8_t
-    _length: [ association_frame_length ] 
+    _type: char
```
I now understand why you make it a `char` instead of `uint8_t` but you must explain that in the commit message - the reason is that currently TLVF has string APIs support for char types so we can easilly print the association frame using `association_frame_str()` which is autogenerated.
However, as I wrote in the previous commit, we must first understand if the association frame should include the null terminator. You can look at the wireshark capture of a passed test with a certified agent to see if it is null terminated.

arnout replied:
> so we can easilly print the association frame using `association_frame_str()`

The association frame is binary data. Printing it as a string will not lead to much joy.

I believe that what you parse out of the AP-STA-CONNECTED event from hostapd is the hex representation of the frame. So it should be converted back to binary (uint8_t) inside bwl and passed as binary in the Client Capability Report TLV.

As Tomer says, check with a certified device what the contents is, and make sure we generate the same thing. Actually, it would be nice if you could upload a pcap file with this TLV as generated by a certified device, so everybody can easily compare with it.

tomereli commented on agent/src/beerocks/slave/ap_manager_thread.cpp:
```diff
@@ -1122,6 +1122,15 @@ bool ap_manager_thread::hal_event_handler(bwl::base_wlan_hal::hal_event_ptr_t ev
+            LOG(DEBUG) << "no association frame";
+        } else {
+            notification->params().association_frame =
+                new char[strnlen(msg->params.association_frame, ASSOCIATION_FRAME_SIZE) + 1];
```
This really looks like a memory leak to me. You allocate a char array - where do you delete it?

tomereli replied:
Also, why do you need to allocate it? you should not do any dynamic allocations when using TLVF, everything is static on the buffer. Please make sure to refactor these parts.

tomereli replied:
Bottom line - you need to change the `sClientAssociationParams::association_frame` from `char *` to an array, and use the same max size used in hostapd.

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -2079,6 +2079,13 @@ bool slave_thread::handle_cmdu_ap_manager_message(Socket *sd,
         notification_out->client_mac() = notification_in->params().mac;
         notification_out->bssid()      = notification_in->params().bssid;
+        if (!notification_in->params().association_frame) {
+            LOG(DEBUG) << "no association frame";
```
The log here is useless and is repeated throughout the entire packet flow, please remove it. It is enough to print once in the origin of the notification and once in the destination, so BWL and controller.

CoralMalachi replied:
I don't think it useless since if the association frame is empty it means something is wrong, and it will be printed only if something wrong..

tomereli replied:
So change to `LOG_IF(!notification_in->params().association_frame, ERROR) << "no association frame";`

arnout replied:
No, it's literally useless, because `association_frame` is an array, not a pointer:

```cpp
#include <iostream>

struct s {
    char foo[10];
};

int main()
{
    s s1;
    s1.foo[0] = '\0';
    if (!s1.foo)
        std::cout << "foo is empty" << std::endl;
    else
        std::cout << "foo is not empty" << std::endl;
}
```

The code above will print "foo is not empty".

It *could* be useful to say `if (notification_in->params().association_frame[0] != '\0')`.

However, as I wrote elsewhere, it should be a binary buffer, not a hex string. IMO.

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -2079,6 +2079,13 @@ bool slave_thread::handle_cmdu_ap_manager_message(Socket *sd,
+        if (!notification_in->params().association_frame) {
+            LOG(DEBUG) << "no association frame";
+        } else {
+            notification_out->set_association_frame(
```
This also makes it clear that the association frame should be a `uint8_t []` and not `char []` in the YAML files.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -245,7 +245,8 @@ class backhaul_manager : public btl::transport_socket_thread {
      * Associated client information is later used to fill in the Associated Clients TLV
      * in the Topology Response message.
      */
```
Please update the comment describing the map.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1735,10 +1735,32 @@ bool backhaul_manager::handle_slave_backhaul_message(std::shared_ptr<SSlaveSocke
+        }
         // Set client association information for associated client
+        std::string assoc_req;
+        if (!msg->association_frame()) {
```
Why not simply leave it as empty string? no need for the debug print and if else.
```suggestion
        std::string assoc_req(msg->association_frame() ? msg->association_frame() : "");
```

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1735,10 +1735,32 @@ bool backhaul_manager::handle_slave_backhaul_message(std::shared_ptr<SSlaveSocke
-        associated_clients[msg->client_mac()] = std::chrono::steady_clock::now();
+
+        associated_clients[msg->client_mac()] =
+            make_tuple(std::chrono::steady_clock::now(), assoc_req);
```
Why use tuple instead of adding a new structure?

```
typedef struct {
    sMacAddr client_mac;
    uint8_t assoc_req[MAX_ASSOC_REQ_SIZE];
    std::chrono::steady_clock::time_point timestamp;
} associated_client_t;

typedef std::unordered_map<sMacAddr, associated_client_t> associated_clients_t;
```

This will make the code much more readable IMO.

CoralMalachi replied:
I have tried to keep the code uniform since we use tuple most of the time, but I like your idea 👍

tomereli replied:
At least I see now that I'm consistent :)
So will you change it?

CoralMalachi replied:
done 👍

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1856,11 +1856,15 @@ bool backhaul_manager::handle_client_capability_query(ieee1905_1::CmduMessageRx
     //with any of the BSS operated by the Multi-AP Agent [ though the TLV does contain a BSSID, the specification
     // says that we should answer if the client is associated with any BSS on this agent.]
     bool associated_client_found = false;
+    sMacAddr client_ruid;
```
Commit message typo: "For that reason the following changes ere made:" --> "For that reason the following changes are made:"

tomereli replied:
So simply add this to the `associated_client_t` struct I suggested in the previous commit:
```
typedef struct {
    sMacAddr client_mac;
    sMacAddr radio_mac; // ruid
    sMacAddr bssid; // VAP mac
    ...
} associated_client_t;
```

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1890,7 +1894,11 @@ bool backhaul_manager::handle_client_capability_query(ieee1905_1::CmduMessageRx
         client_capability_report_tlv->result_code() = wfa_map::tlvClientCapabilityReport::SUCCESS;
         LOG(DEBUG) << "Result Code: SUCCESS";
-        //TODO: Add frame body of the most recently received (Re)Association Request frame from this client
+        //Add frame body of the most recently received (Re)Association Request frame from this client
```
Add space after `//` (nitpick)

tomereli commented on common/beerocks/bwl/dummy/ap_wlan_hal_dummy.cpp:
```diff
@@ -376,9 +376,12 @@ bool ap_wlan_hal_dummy::process_dummy_event(parsed_obj_map_t &parsed_obj)
             LOG(ERROR) << "Failed reading mac parameter!";
             return false;
         }
+        msg->params.association_frame = new char[30];
```
Commit message should explain why we add a dummy association frame. If you do that as a preparation for a test (and you should) - then mention that.

CoralMalachi replied:
Yes I know it was not ready yet :)

tomereli reviewed (changes requested):
As usual, great job with the commit split and documentation. Your PR description is a sight to behold, I used it as an example for new contributors joining prplMesh :)

However, I have quite a lot of comments, some require big changes so requesting changes for now.
Please also verify this in the testbed so we know that with this we pass the full tests.

morantr commented on framework/tlvf/AutoGenerated/include/tlvf/wfa_map/tlvClientCapabilityReport.h:
```diff
@@ -22,6 +22,7 @@
 #include <tlvf/ClassList.h>
```
This commit (5ac1d342585845243a90ea271f7b31458260abce : tlvf: update cliaent caps report TLV auto-generated files) has a typo on the headline.
cliaent ->client

morantr commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -146,6 +146,9 @@ sClientAssociationParams:
   bssid: sMacAddr
```
This commit message headline (ad65f91c5bc80152c1fb29449e2795d4dfc2a58e tlvf: beerocks message common: add association frame to sClientAssoci…) is too long (>72 characteres).

morantr commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -146,6 +146,9 @@ sClientAssociationParams:
   capabilities: beerocks::message::sRadioCapabilities
   vap_id: int8_t
+  association_frame:
+    _type: char
```
Why char and not uint8_t?

tomereli replied:
+1. we should use `uint8_t` everywhere now for the association frame, cause its binary data.

morantr commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml:
```diff
@@ -125,6 +125,9 @@ cACTION_BACKHAUL_CLIENT_ASSOCIATED_NOTIFICATION:
   iface_mac: sMacAddr
```
This commit message (cdbaca92af7d91485d92f7b05bc8b2fc7e768e62: tlvf: beerocks message backhaul: add association frame to CLIENT_ASSO…) is too long.
You can remove "beerocks message backhaul" and just write "tlvf:".

morantr commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2093,7 +2093,7 @@ bool backhaul_manager::handle_1905_topology_query(ieee1905_1::CmduMessageRx &cmd
                     // the timestamp of its last association.
```
commit message (f72a6ab22aaaf49cbf4dddd9da7d1e0d24d232b7: agent: backhaul manager: update db to save association frame for each…) headline is too long.

morantr commented on common/beerocks/bwl/dummy/ap_wlan_hal_dummy.cpp:
```diff
@@ -376,9 +376,12 @@ bool ap_wlan_hal_dummy::process_dummy_event(parsed_obj_map_t &parsed_obj)
         }
+        msg->params.association_frame = new char[30];
+        msg->params.mac               = beerocks::net::network_utils::mac_from_string(tmp_str);
+        char assoc_req[]              = "dummy association frame";
```
Not sure why it is required to write `"dummy association frame"` on the tlv, but anyway it probably will better if you use std::string instead of c-string.

morantr reviewed (approved):
Besides Tomer's comments that I agree with, I added a few small comments.
Anyway, I'll approve it so you will not be blocked by me.

Commit:
common: beerocks: add association length constant

The absolute maximum size of any frame according to the 802.11
specifications is MMPDU size of 2304 bytes. The actual size of a
(re)association frame should be in the range of a couple of hundreds of
bytes, so be on the safe side and set the maximum size to 2KB.

For that reason, add a constant variable that is set to the desired size.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
bwl: dwpal: AP-STA-CONNECTED parse assoc_req

Preparative commit.
Since EasyMesh client capability report (4.6.2) requires the client
(re)association frame, parse it from the AP-STA-CONNECTED event.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>
Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
controller: log the association frame if it received

Preparative commit.
if this is no error scenario, it means that the
CLIENT_CAPABILITY_REPORT_MESSAGE should contain the most recently
received (Re)Association Request frame from the client.
Log the (Re)Association frame to validate the flow succeeded in the test
flows later.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

arnout commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -43,6 +43,7 @@ enum eStructsConsts {
     HOSTAP_ERR_MSG_LENGTH         = 64,
     WIFI_DRIVER_VER_LENGTH        = 32 + 4,
     WIFI_SSID_MAX_LENGTH          = 32 + 1 + 3, //need extra 1 byte for null termination + alignment
+    ASSOCIATION_MAX_LENGTH        = 2048,
```
I really would like some explanation (in the commit message or as a comment) about how you get this number.

tomereli replied:
I was sure I explained this in the commit which changed this originally.
I don't remember the maximum size - MPDU length, I think its ~2300 Bytes.
According to the FW guys, the association frame is normally a couple of hundreds of bytes so 2KB should be well enough.

CoralMalachi replied:
I thought it would be nicer to add a constant variable that is set to the desired size. I can move your explained to the place where ASSOCIATION_MAX_LENGTH  is defined (beerocks_defines.h). @tomereli :)

tomereli replied:
Please also add the explanation to the commit message, don't leave it empty.

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -1812,15 +1812,16 @@ bool ap_wlan_hal_dwpal::process_dwpal_event(char *buffer, int bufLen, const std:
+        char ht_cap[8]                         = {0};
+        char vht_cap[16]                       = {0};
+        size_t numOfValidArgs[20]              = {0};
+        char assoc_req[ASSOCIATION_FRAME_SIZE] = {0};
```
Note that *here*, char is the correct type.

However, this size should be `2*ASSOCIATION_FRAME_SIZE+1` to fix the hex representation + the terminating \0.

CoralMalachi replied:
why 2 times ASSOCIATION_FRAME_SIZE?

arnout replied:
Because the 2048 was assumed to be the maximum of the *binary* frame, but it's represented as hex here. So you need two hex characters per byte + a terminating \0.

arnout commented on common/beerocks/bwl/include/bwl/base_wlan_hal_types.h:
```diff
@@ -252,6 +252,11 @@ typedef struct {
+// The absolute maximum size of any frame according to the 802.11 specification
+// is MMPDU size of 2304 bytes. The actual size of an (re)association frame
+// should be in the range of a couple of hundreds of bytes, so be on the safe side
+// and set the maximum size to 2KB
```
Move this explanation to the place where the number 2048 is used.

tomereli replied:
Oh here it is!

CoralMalachi replied:
it used here - ASSOCIATION_MAX_LENGTH is defined to 2048 in beerocks_defines.h, do you want me to move this comment there?

arnout replied:
Yes, that's what we mean.

The comment explains why the value 2048 is chosen, but that value is not defined here, it's defined in beerocks_defines.h.

CoralMalachi replied:
done

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -1902,6 +1902,8 @@ bool ap_wlan_hal_dwpal::process_dwpal_event(char *buffer, int bufLen, const std:
         msg->params.mac          = beerocks::net::network_utils::mac_from_string(MACAddress);
         msg->params.capabilities = {};
+        std::copy_n(assoc_req, strnlen(assoc_req, ASSOCIATION_FRAME_SIZE) + 1,
+                    msg->params.association_frame);
```
As I note elsewhere: you shouldn't simply copy, but convert to binary.

CoralMalachi replied:
I changed it but did not push it yet since I want to validate it works on testbed first (but I can't since the master branch using it now).

Currently, I'm facing an issue in which the test sometimes passes and sometimes not, without changing a thing (on a side branch).
I'll try to convert as you suggested in the ap_wlan_hal_dwpal.cpp (currently I'm converting it only before sending it as part of the client capability report message and maybe it's wrong..)

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -146,6 +146,9 @@ sClientAssociationParams:
   vap_id: int8_t
+  association_frame:
+    _type: char
+    _length: [ "beerocks::message::ASSOCIATION_MAX_LENGTH" ] 
```
You really should use a dynamic length here, because there is no 0 terminator to indicate the length. But then it can't be a struct any more :-( But actually, this struct really serves no purpose at all, it can be inlined directly into cACTION_APMANAGER_CLIENT_ASSOCIATED_NOTIFICATION. So please do that (in a separate commit) and then add the association frame as a dynamic length field.

Note that in the 1905 message it's an unknown length field, but that's not possible here because we don't have a total length in the beerocks classes.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -2105,6 +2105,14 @@ bool slave_thread::handle_cmdu_ap_manager_message(Socket *sd,
+            notification_out->set_association_frame(
+                notification_in->params().association_frame,
+                strnlen(notification_in->params().association_frame, ASSOCIATION_FRAME_SIZE) + 1);
+        }
```
Meh I hate all this useless copying around of notification messages...

CoralMalachi replied:
Changed it.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1742,10 +1742,25 @@ bool backhaul_manager::handle_slave_backhaul_message(std::shared_ptr<SSlaveSocke
+        //remove this client from other radios
+        for (const auto &slave : m_radio_info_map) {
+            auto associated_clients_map = slave.second.associated_clients_map;
+            for (const auto &vap : associated_clients_map) {
```
vap? These are associated clients, right? So `client` would be a better name.

CoralMalachi replied:
Done 👍

arnout commented on common/beerocks/bwl/dummy/ap_wlan_hal_dummy.cpp:
```diff
@@ -379,9 +379,11 @@ bool ap_wlan_hal_dummy::process_dummy_event(parsed_obj_map_t &parsed_obj)
             return false;
         }
+        msg->params.mac  = beerocks::net::network_utils::mac_from_string(tmp_str);
+        char assoc_req[] = "dummy association frame";
```
I think it would be better to include a real association frame here, e.g. something that you got from sniffing actual easymesh TLVs from a certified device.

CoralMalachi replied:
done.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1022,6 +1022,10 @@ bool master_thread::handle_cmdu_1905_client_capability_report_message(
               << ", client MAC= " << network_utils::mac_to_string(client_info_tlv->client_mac())
               << ", BSSID= " << network_utils::mac_to_string(client_info_tlv->bssid());
+
+    LOG_IF(!result_code.compare("SUCCESS"), DEBUG)
```
(nitpick) `result_code == "SUCCESS"` is more natural to me.

Or actually, `client_capability_report_tlv->result_code() == wfa_map::tlvClientCapabilityReport::SUCCESS` is even better.

tomereli replied:
Shouldn't really be a nitpick - its in our CONTRIBUTING.md AFAIR.

arnout commented on tests/test_flows.sh:
```diff
@@ -197,6 +197,7 @@ test_client_capability_query() {
     dbg "Confirming client capability report message has been received on controller"
     check_log ${GATEWAY} controller "Received CLIENT_CAPABILITY_REPORT_MESSAGE"
     check_log ${GATEWAY} controller "Result Code= SUCCESS, client MAC= ${sta_mac2}, BSSID= ${mac_agent1_wlan0}"
+    check_log ${GATEWAY} controller "(Re)Association Request frame= dummy association frame"
```
Please update test_flows.py instead.

CoralMalachi replied:
done

arnout reviewed (approved):
The confusion between bytes and hex is a major concern, but can be fixed easily.

Commit:
common: hex to char method

Preparative commit.
add hex_to_char_string method which will be used for converting the
string we get from the DWPAL from hex format fo char ASCII format.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
agent: client caps report: add most recently received (Re)Association frame

According to Multi-AP Specification, If a Multi-AP Agent receives a
Client Capability Query message, then within one second it shall respond
with a Client Capability Report message.

If the STA specified in the Client Capability Query message is not
associated with any of the BSS operated by the Multi-AP Agent (an error
scenario), the Multi-AP Agent shall set the result code in the Client
Capability Report TLV to 0x01 and include an Error Code TLV.
If it is not an error scenario, the Client Capability Report TLV should
include the frame body of the most recently received (Re)Association
Request frame from this client.

For that reason the following changes ere made:
1. When searching for the STA specified in the Client Capability Query in
   the backhaul manager map, save the RUID and the VAP that the client is
   associated with.

2. Now, we can easily get from the backhaul map the frame body of the most
   recently received (Re)Association Request frame from this client and add
   it to the Client Capability Report TLV.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

tomereli commented on common/beerocks/bwl/include/bwl/base_wlan_hal_types.h:
```diff
@@ -190,6 +190,7 @@ typedef struct {
     uint8_t reserved1;
     uint8_t reserved2;
     uint8_t reserved3;
+    char association_frame[beerocks::message::ASSOCIATION_MAX_LENGTH];
```
Better to use `uint8_t` which makes it clear that this is binary data.

CoralMalachi replied:
done.

tomereli commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -1902,6 +1902,10 @@ bool ap_wlan_hal_dwpal::process_dwpal_event(char *buffer, int bufLen, const std:
         msg->params.vap_id       = beerocks::utils::get_ids_from_iface_string(VAP).vap_id;
```
Commit message "the assoc_req we get from the DWPAL the whole message" --> "the assoc_req we get from the DWPAL **contains** the whole message.
Please also add an explanation in the commit message and in the code - nice big comment - with reference to where in the specifications it says that the association request should be only the tagged parameters, that is if you know it. If not - perhaps @arnout does?

CoralMalachi replied:
fixed the commit message.

tomereli commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -1902,6 +1902,10 @@ bool ap_wlan_hal_dwpal::process_dwpal_event(char *buffer, int bufLen, const std:
         msg->params.mac          = beerocks::net::network_utils::mac_from_string(MACAddress);
         msg->params.capabilities = {};
+        auto assoc_frame_len     = strnlen(assoc_req, ASSOCIATION_FRAME_SIZE);
+        auto sub_str             = std::string(&assoc_req[56], assoc_frame_len - 56 - 18);
```
Don't use magic numbers, please.  Why 56? why 18? These needs to be defined or at least documented in a comment with a TODO perhaps saying how this should be done properly.

CoralMalachi replied:
I explained it in the commit message... I'll add a comment in the code as well.

arnout replied:
It should still be defined with static const variables.

CoralMalachi replied:
Ok 👍 @arnout

tomereli commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -1902,6 +1902,10 @@ bool ap_wlan_hal_dwpal::process_dwpal_event(char *buffer, int bufLen, const std:
+        auto assoc_frame_len     = strnlen(assoc_req, ASSOCIATION_FRAME_SIZE);
+        auto sub_str             = std::string(&assoc_req[56], assoc_frame_len - 56 - 18);
+
+        std::copy_n(&sub_str[0], sub_str.length() + 1, msg->params.association_frame);
```
why `&sub_str[0]` and not just `sub_str`? Why the extra `1`? null termination is not required IMO.

CoralMalachi replied:
If I pass just `sub_str` it does not compile...
about the extra byte, I removed it.

tomereli commented on framework/tlvf/yaml/tlvf/wfa_map/tlvClientCapabilityReport.yaml:
```diff
@@ -10,12 +10,9 @@ tlvClientCapabilityReport:
   association_frame:
     _type: uint8_t
-    _length: [ association_frame_length ] 
+    _length: [] 
```
Commit message is wrong - the type is `uint8_t` (as it should be) and not `char`.

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml:
```diff
@@ -125,6 +125,9 @@ cACTION_BACKHAUL_CLIENT_ASSOCIATED_NOTIFICATION:
   client_mac: sMacAddr
   bssid: sMacAddr
+  association_frame:
+    _type: char 
```
type should be `uint8_t`.

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml:
```diff
@@ -125,6 +125,9 @@ cACTION_BACKHAUL_CLIENT_ASSOCIATED_NOTIFICATION:
   bssid: sMacAddr
+  association_frame:
+    _type: char 
+    _length: [] 
```
type should be `uint8_t`.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -246,7 +246,8 @@ class backhaul_manager : public btl::transport_socket_thread {
-    typedef std::unordered_map<sMacAddr, std::chrono::steady_clock::time_point>
+    typedef std::unordered_map<sMacAddr,
+                               std::tuple<std::chrono::steady_clock::time_point, std::string>>
         associated_clients_t;
```
```suggestion

struct sClientInfo {
    sMacAddr client_mac;
    std::chrono::steady_clock::time_point;
    uint8_t assoc_req[MAX_ASSOC_REQ_LEN];
};

typedef std::unordered_map<sMacAddr, sClientInfo> associated_clients_t;
```

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1742,10 +1742,25 @@ bool backhaul_manager::handle_slave_backhaul_message(std::shared_ptr<SSlaveSocke
             return false;
         }
 
+        //remove this client from other radios
```
Please move to a private helper function (like I did here https://github.com/prplfoundation/prplMesh/pull/1067/commits/f52ed42dee312e662cdfac61c114a3178ff1c018)

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1742,10 +1742,25 @@ bool backhaul_manager::handle_slave_backhaul_message(std::shared_ptr<SSlaveSocke
+            }
+        }
         // Set client association information for associated client
+        std::string assoc_req(msg->association_frame() ? msg->association_frame() : "");
```
Should be `uint8_t`

tomereli commented on common/beerocks/bwl/dummy/ap_wlan_hal_dummy.cpp:
```diff
@@ -379,9 +379,19 @@ bool ap_wlan_hal_dummy::process_dummy_event(parsed_obj_map_t &parsed_obj)
-
         msg->params.mac = beerocks::net::network_utils::mac_from_string(tmp_str);
-
+        char assoc_req[] =
```
`const char`

CoralMalachi replied:
done

tomereli commented on common/beerocks/bwl/dummy/ap_wlan_hal_dummy.cpp:
```diff
@@ -379,9 +379,19 @@ bool ap_wlan_hal_dummy::process_dummy_event(parsed_obj_map_t &parsed_obj)
+            "0103";
+
+        auto assoc_req_len = strnlen(assoc_req, ASSOCIATION_FRAME_SIZE);
+        auto sub_str       = std::string(&assoc_req[56], assoc_req_len - 56 - 18);
```
Same thing - please no magic numbers.

tomereli commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -1836,6 +1837,8 @@ bool ap_wlan_hal_dwpal::process_dwpal_event(char *buffer, int bufLen, const std:
              "VHT_MCS=", sizeof(VHT_MCS)},
             {(void *)ht_cap, &numOfValidArgs[7], DWPAL_STR_PARAM, "HT_CAP=", sizeof(ht_cap)},
             {(void *)vht_cap, &numOfValidArgs[8], DWPAL_STR_PARAM, "VHT_CAP=", sizeof(vht_cap)},
+            {(void *)assoc_req, &numOfValidArgs[9], DWPAL_STR_PARAM,
```
Maybe it is better to do the conversion to the offset and size we need already here (nitpick)

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1022,6 +1022,13 @@ bool master_thread::handle_cmdu_1905_client_capability_report_message(
+               wfa_map::tlvClientCapabilityReport::SUCCESS,
+           DEBUG)
+        << "(Re)Association Request frame= ";
+
```
You can just use `utils::dump_buffer` - no need to convert to ascii.

CoralMalachi replied:
thanks

tomereli reviewed (changes requested):
There is a big mess in the commits. The assoc_req needs to be uint8_t - so asking changes for now.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2105,11 +2119,18 @@ bool backhaul_manager::handle_client_capability_query(ieee1905_1::CmduMessageRx
+    // sMacAddr client_ruid;
+    // sMacAddr client_vap;
     for (const auto &slave : m_radio_info_map) {
         auto associated_clients_map = slave.second.associated_clients_map;
```
```suggestion
        const auto &associated_clients_map = slave.second.associated_clients_map;
```

This was already wrong before this PR, but I notice it now...

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2135,18 +2156,35 @@ bool backhaul_manager::handle_client_capability_query(ieee1905_1::CmduMessageRx
+        client_capability_report_tlv->alloc_association_frame(s.asso_len / 2);
+        std::copy(&assoc_frame_body[0], &assoc_frame_body[0] + s.asso_len / 2,
+                  client_capability_report_tlv->association_frame());
+
```
This whole thing is wrong, no string conversion is needed here.

```suggestion
        client_capability_report_tlv->alloc_association_frame(s.asso_len);
        std::copy_n(s.assoc_req, s.asso_len, client_capability_report_tlv->association_frame());
```

CoralMalachi replied:
sorry, the branch was not updated, that's why I named the commit message WIP :)
I've pushed now the side branch, and I'll go over all your comments @arnout, thanks.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -242,21 +244,33 @@ class backhaul_manager : public btl::transport_socket_thread {
+        sMacAddr radio_mac; // ruid
+        sMacAddr bssid;     // VAP mac
+        std::chrono::steady_clock::time_point time_stamp;
+        size_t asso_len;
+        uint8_t assoc_req[ASSOCIATION_FRAME_SIZE];
```
```suggestion
        std::array<uint8_t> assoc_req;
```

If you see a size and array combination in any structure that is not tlv-generated, it's probably wrong :-). Actually, if you see a C-style array it's probably wrong.

Hm, maybe I need to make a CONTRIBUTING.md update...

Note that with the std::array, the copy is simply `std::copy(s.assoc_req, client_capability_report_tlv->association_frame())`.

CoralMalachi replied:
I agree it does not make sense to use size and array combination inside a structure.
The reason I did it is that I need to know the length of each association from so when I pass it in the client capability report it will have the correct length.

One option I thought about is to use a vector, which is dynamic but it isn't very comfortable to use (to push every time another `uint8_t`).
Another is to declare it as c array and to fill it with a null terminator (as we did in the ap_wlan_hal Event::STA_Connected) and then we can use `strnlen` after converting to string.

but `std::array` also requires a length argument so it's also not a dynamic array.
I want to know the length of each association frame and not take the fixed size, so how does `std::array` help me here? let me know if I did not understand your suggestion @arnout

arnout replied:
OK, I was mixing things up a little, sorry about that...

`uint8_t assoc_req[ASSOCIATION_FRAME_SIZE];` is equivalent to `std::array<uint8_t, ASSOCIATION_FRAME_SIZE> assoc_req;` - but the latter would still require asso_len. So in this case, it should indeed be a `std::vector`.

However, as you observe, a vector is not very comfortable, because we later want to copy it out with set_association_frame(). Well, actually, it's not that bad, because we can use `std::copy_n` to fill in the vector, and `std::vector::data()` to access the data directly for set_association_frame().

However, as I indicated before, I think it's better to use a `basic_string<uint8_t>` for such a situation.

I'm working on a quick commit to show what I mean.

arnout replied:
I pushed my proposal to dev/client-caps-report-assoc_req-with-bytes (additional fixup commits on top of this PR). If it looks good to @CoralMalachi and @tomereli just take those commits in this PR and rebase on master. I didn't create a PR for that branch.

Note that this is not terribly important because we already use this C-style array pattern in dozens of places in the code, so adding one more is not a huge concern. And we really should get this PR merged soonish, it's been dragging on for too long..

tomereli replied:
I like it. I basic_string is a weird name for storing binary data so using a bytes class is welcome. I now understand why @mariomaz used std::basic_string in his PR - so this should probably be pushed as a separate PR so we can merge it quickly - @arnout can you do that so @mariomaz can rebase his PR and use the `bytes` class?

Commit:
bwl: dwpal: add association frame to sClientAssociationParams

Preparative commit.
Add an association frame to sClientAssociationParams struct.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

arnout commented on common/beerocks/bwl/dummy/ap_wlan_hal_dummy.cpp:
```diff
@@ -380,9 +380,27 @@ bool ap_wlan_hal_dummy::process_dummy_event(parsed_obj_map_t &parsed_obj)
+        LOG(DEBUG)
+            << "*********************************************************************************";
+
+        std::copy_n(&sub_str[0], sub_str.length(), msg->params.association_frame);
```
The association_frame in the output is binary, not hex, so you have to convert the hex string to binary first.

CoralMalachi replied:
done.

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -1868,6 +1872,11 @@ bool ap_wlan_hal_dwpal::process_dwpal_event(char *buffer, int bufLen, const std:
+        auto sub_str             = std::string(&assoc_req[56], assoc_frame_len - 56 - 18);
+        msg->params.association_frame_length = sub_str.length();
+
+        std::copy_n(&sub_str[0], sub_str.length() + 1, msg->params.association_frame);
```
Again, it should already have been converted to binary here.

I realize I said before that you could choose to pass it around as string internally, but really there is no reason to do that, so let's just do the right thing and convert to binary here.

CoralMalachi replied:
done.

arnout commented on tests/test_flows.py:
```diff
@@ -466,6 +479,9 @@ def test_client_capability_query(self):
                        .format(sta2.mac, env.agents[0].radios[0].mac))
+               
+        for line in association_frame.splitlines():
+            self.check_log(env.controller, r"{}".format(line))
```
Can you use the start_line feature that was recently added to check_log to make sure that you're actually matching the right part of the log? The previous check_log call just before the loop will give you the starting line.

CoralMalachi replied:
Do you mean instead of the loop or just before that? I don't think I understood the purpose of this method.

arnout replied:
Never mind, I'll just rework the startline functionality entirely at some point.

Commit:
tlvf: inlined the struct directly into the class

Preparative commit.

Since this struct really serves no purpose at all, it can be inlined directly into cACTION_APMANAGER_CLIENT_ASSOCIATED_NOTIFICATION.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: add autogenerated files

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: add the association frame as a dynamic length field

Preparative commit.

When the AP manager received the Event:: STA_Connected, a vendor-specific
message of type `cACTION_APMANAGER_CLIENT_ASSOCIATED_NOTIFICATION` is
created and being sent to the son slave.
In order to add to this message the association frame, we need to add a
new field to cACTION_APMANAGER_CLIENT_ASSOCIATED_NOTIFICATION of type uint8_t.

We should use a dynamic length here because there is no 0 terminator to
indicate the length, and that is the reason why we couldn't use the
struct anymore.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: beerocks message apm: add the autogenerated files

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
agent: backhaul manager: update db to save association frame for each client

save in the associated_clients_t map the frame body of the most recently
received (Re)Association Request frame from each client.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
agent: backhaul manager: add remove_client_from_all_radios method

Preparative commit.

add a method that is responsible for removing a client from other radios.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
agent: backhaul manager: save association frame by client

When the ACTION_BACKHAUL_CLIENT_ASSOCIATED_NOTIFICATION vendor-specific
message is received on the backhaul manager,  remove that client form the
associated_clients_t map (since maybe that client was associated with
another radio) and fill the info again with the frame body of the most
recently received (Re)Association Request frame from this client.

That way the associated_clients_t map always contains info of each client,
and we can easily check which radio and VAP the client is associated with
and get the association frame.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
test_flows: update client capability test

add a check in the `test_client_capability_query` which validates that
the association frame was passed correctly from the ap_wlan_hal --->
ap_manager ---> son_slave ---> backhaul_manager ---> controller.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
agent: backhaul manager: add sClientInfo

Preparative commit.

For convenience, we will keep in a single structure all the information
we want to have about a client.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
bwl: dwpal: send association frame to ap manager

Preparative commit.
Send the association frame to the AP manager, as part of the
STA_Connected event.

As mentioned in the previous commit message, the association frame we get
from the DWPAL contains the whole message, but the client capability
report should include only the frame body of the association request.

It is not clear from the specification what are the exact fields of the
message should be passed, but since certified units are passing only the
tagged parameters (after cleaning the Wi-Fi Alliance tag at from the end),
we do the same and pass the tagged params only. Otherwise, if we pass the
whole frame, the tests fail.

Therefore, we send only a substring of the original association frame we
get from the DWPAL, according to the substring range which was defined in
the previous commit.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
common: ap_wlan_hal: add static const variables

Preparative commit.
In the next steps, we will send the association frame to the AP manager,
as part of the STA_Connected event.

Please note, the association frame we get from the DWPAL contains the
whole message, but the client capability report should include only the
frame body of the association request.

It is not clear from the specification what are the exact fields of the
message should be passed, but since certified units are passing only the
tagged parameters (after cleaning the Wi-Fi Alliance tag at from the end),
we do the same and pass the tagged params only. Otherwise, if we pass the
whole frame, the tests fail.

Therefore, the range of the subframe we want to send are defined as
static const variables.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
common: ap_wlan_hal: get binary association frame method

add a method that fetches the substring of the original association frame we
get from the DWPAL, according to the substring range which was defined in
the previous commit, and then convert it to binary form.

the method is defined as a protected method so it
can be accessed in the class and its derived classes ( ap_wlan_hal_dwpal
& ap_wlan_hal_dummy)

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
sClientAssociationParams: add bssid field

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
rdkb_wlan_task: use sClientAssociationParams

Signed-off-by: Coral Malachi <coral.malachi@intel.com>
